### PR TITLE
async publication to kafka

### LINKED ISSUE #744: when kafka is not responding client is blocked until timeout 60s
For some reason kafka was not reponding, request on api-server to delete a nuvlabox took arount 60s to complete because kafka client was waiting timeout

### PR DESCRIPTION
**Pull request by konstan**
- implements async publication of the messages from the server resources to Kafka.

Future improvements:

- #764 
- #765

Fixes #744

### CODE REVIEW

File: code/src/sixsq/nuvla/server/util/kafka.clj [hunk 1]
```diff
@@ -37,52 +63,143 @@
...
-    (log/info "publish test message to" topic "by:" producer)
-    (publish topic nil {:producer (str producer)})))
 
+(defn chan-consumer-kafka-producer
+  [id kafka-producer]
+  (a/go
+    (loop []
+      (when-let [{:keys [topic key value] :as msg} (<! *comm-chan*)]
+        (log/debugf "channel consumer %s consumed: %s" id msg)
+        (kc/send! kafka-producer topic key value)
```
Comment (konstan) [inline code/src/sixsq/nuvla/server/util/kafka.clj:123]: Note for me: try/catch would be useful here.

File: code/src/sixsq/nuvla/server/util/kafka.clj [hunk 2]
```diff
@@ -1,24 +1,93 @@
...
+; Communication channel.
+(def comm-chan! (atom (when (need-init?)
+                        (let [len (comm-channel-len-from-env)]
+                          (log/info "initialise async comm channel of len:" len)
+                          (chan (a/sliding-buffer len))))))
+
+
+(defn comm-chan-set!
+  [len]
+  (swap! comm-chan! (constantly (chan (a/sliding-buffer len)))))
```
Comment (0xbase12) [inline code/src/sixsq/nuvla/server/util/kafka.clj:77]: `(reset! comm-chan! (chan (a/sliding-buffer len))`

File: code/src/sixsq/nuvla/server/util/kafka.clj [hunk 3]
```diff
@@ -37,52 +106,102 @@
...
+  [id producer]
+  (a/go-loop []
+    (if (some? @comm-chan!)
+      (when-let [{:keys [topic key value] :as msg} (<! @comm-chan!)]
+        (log/debugf "producer %s consumed from comm chan: %s" id msg)
+        (try
+          (kc/send! producer topic key value)
+          (log/debugf "producer %s published: %s %s %s" id topic key value)
+          (catch Exception e
+            (log/error "producer %s %s failed publishing to kafka: %s" id producer e)))
```
Comment (0xbase12) [inline code/src/sixsq/nuvla/server/util/kafka.clj:150]: log/errorf

File: code/src/sixsq/nuvla/server/util/kafka.clj [hunk 4]
```diff
@@ -37,52 +106,102 @@
...
+        (recur))
+      (log/warnf "comm channel is not defined"))))
```
Comment (0xbase12) [inline code/src/sixsq/nuvla/server/util/kafka.clj:152]: log/warn

File: code/src/sixsq/nuvla/server/util/kafka.clj [hunk 5]
```diff
@@ -37,52 +106,102 @@
...
+  []
+  (when (pos? (count @producers!))
+    (log/debugf "closing kafka producers: %s" @producers!)
+    (doseq [[id p] @producers!]
+      (log/debugf "closing producer %s: %s" id p)
+      (try
+        (kc/close! p)
+        (catch Exception e
+          (log/errorf "failed closing producer %s %s: %s" id p e))))
+    (swap! producers! (constantly {}))))
```
Comment (0xbase12) [inline code/src/sixsq/nuvla/server/util/kafka.clj:193]: `(reset! producers! {})`

File: code/src/sixsq/nuvla/server/util/kafka.clj [hunk 6]
```diff
@@ -37,52 +106,102 @@
...
+    (start-producers! (producers-num-from-env)
+                      (or bootstrap-servers (bootstrap-servers-from-env)))
+    (log/warn "kafka producers will not be created.")))
+
+
+(defn close-producers!
+  "Closes all Kafka producers and destroys their communication channel with the
+  client code."
+  []
+  (when (pos? (count @producers!))
```
Comment (0xbase12) [inline code/src/sixsq/nuvla/server/util/kafka.clj:185]: `(when (seq @producers!)`

Comment (konstan) [reply code/src/sixsq/nuvla/server/util/kafka.clj]: addressed

File: code/src/sixsq/nuvla/server/util/kafka.clj (hunk 5 above)
Comment (konstan) [reply code/src/sixsq/nuvla/server/util/kafka.clj]: addressed

File: code/src/sixsq/nuvla/server/util/kafka.clj (hunk 4 above)
Comment (konstan) [reply code/src/sixsq/nuvla/server/util/kafka.clj]: addressed

File: code/src/sixsq/nuvla/server/util/kafka.clj (hunk 2 above)
Comment (konstan) [reply code/src/sixsq/nuvla/server/util/kafka.clj]: addressed

File: code/src/sixsq/nuvla/server/util/kafka.clj [hunk 7]
```diff
@@ -37,52 +63,143 @@
...
-  [producer]
-  (log/info "setting new kafka producer:" producer)
-  (alter-var-root #'*producer* (constantly producer))
-  (let [topic "nuvla-kafka"]
-    (log/info "publish test message to" topic "by:" producer)
-    (publish topic nil {:producer (str producer)})))
 
+(defn chan-consumer-kafka-producer
+  [id kafka-producer]
+  (a/go
```
[the remark this answers is not in the document]
Comment (konstan) [reply code/src/sixsq/nuvla/server/util/kafka.clj]: addressed

**Pull request merged**